# Assignment 5 - Text Analysis
An explanation this assignment could be found in the .pdf explanation document


## Materials to review for this assignment
<h4>From Moodle:</h4> 
<h5><u>Review the notebooks regarding the following python topics</u>:</h5>
<div class="alert alert-info">
&#x2714; <b>Working with strings</b> (tutorial notebook)<br/>
&#x2714; <b>Text Analysis</b> (tutorial notebook)<br/>
&#x2714; <b>Hebrew text analysis tools (tokenizer, wordnet)</b> (moodle example)<br/>
&#x2714; <b>(brief review) All previous notebooks</b><br/>
</div> 
<h5><u>Review the presentations regarding the following topics</u>:</h5>
<div class="alert alert-info">
&#x2714; <b>Text Analysis</b> (lecture presentation)<br/>
&#x2714; <b>(brief review) All other presentations</b><br/>
</div>

## Preceding Step - import modules (packages)
This step is necessary in order to use external modules (packages). <br/>

In [1]:
# --------------------------------------
import pandas as pd
import numpy as np
# --------------------------------------


# --------------------------------------
# ------------- visualizations:
import seaborn as sns
import matplotlib.pyplot as plt
from matplotlib.colors import ListedColormap
# --------------------------------------


# ---------------------------------------
import sklearn
from sklearn import preprocessing, metrics, pipeline, model_selection, feature_extraction 
from sklearn import naive_bayes, linear_model, svm, neural_network, neighbors, tree
from sklearn import decomposition, cluster

from sklearn.model_selection import train_test_split, cross_val_score, GridSearchCV 
from sklearn.pipeline import Pipeline
from sklearn.metrics import accuracy_score, confusion_matrix
from sklearn.metrics import precision_score, recall_score, f1_score
from sklearn.metrics import mean_squared_error, r2_score, silhouette_score
from sklearn.preprocessing import MinMaxScaler, StandardScaler, LabelEncoder

from sklearn.svm import LinearSVC
from sklearn.neural_network import MLPClassifier
from sklearn.linear_model import Perceptron, SGDClassifier
from sklearn.decomposition import PCA
from sklearn.cluster import KMeans
from sklearn.naive_bayes import MultinomialNB, GaussianNB
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
# ---------------------------------------


# ----------------- output and visualizations: 
import warnings
from sklearn.exceptions import ConvergenceWarning
warnings.simplefilter("ignore")
warnings.simplefilter(action='ignore', category=FutureWarning)
warnings.simplefilter("ignore", category=ConvergenceWarning)
# show several prints in one cell. This will allow us to condence every trick in one cell.
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"
%matplotlib inline
pd.pandas.set_option('display.max_columns', None)
pd.set_option('display.float_format', lambda x: '%.3f' % x)
# ---------------------------------------

### Text analysis and String manipulation imports:

In [2]:
# --------------------------------------
# --------- Text analysis and Hebrew text analysis imports:
# vectorizers:
from sklearn.feature_extraction import text
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer

# regular expressions:
import re
# --------------------------------------

### (optional) Hebrew text analysis - WordNet (for Hebrew)
Note: the WordNet is not a must

#### (optional) Only if you didn't install Wordnet (for Hebrew) use:

In [ ]:
# word net installation:

# unmark if you want to use and need to install
# !pip install wn
# !python -m wn download omw-he:1.4

In [ ]:
# word net import:

# unmark if you want to use:
# import wn

### (optional) Hebrew text analysis - hebrew_tokenizer (Tokenizer for Hebrew)
Note: the hebrew_tokenizer is not a must

#### (optional) Only if you didn't install hebrew_tokenizer use:

In [ ]:
# Hebrew tokenizer installation:

# unmark if you want to use and need to install:
# !pip install hebrew_tokenizer

In [ ]:
# Hebrew tokenizer import:

# unmark if you want to use:
# import hebrew_tokenizer as ht

### Reading input files
Reading input files for train annotated corpus (raw text data) corpus and for the test corpus

In [3]:
train_filename = 'annotated_corpus_for_train.csv'
test_filename  = 'corpus_for_test.csv'
df_train = pd.read_csv(train_filename, index_col=None, encoding='utf-8')
df_test  = pd.read_csv(test_filename, index_col=None, encoding='utf-8')

In [4]:
df_train.head(8)
df_train.shape

story gender
0  כשחבר הזמין אותי לחול, לא באמת חשבתי שזה יקרה,...      m
1  לפני שהתגייסתי לצבא עשיתי כל מני מיונים ליחידו...      m
2  מאז שהתחילו הלימודים חלומו של כל סטודנט זה הפנ...      f
3  כשהייתי ילד, מטוסים היה הדבר שהכי ריתק אותי. ב...      m
4  ‏הייתי מדריכה בכפר נוער ומתאם הכפר היינו צריכי...      f
5  לפני כ3 חודשים טסתי לרומא למשך שבוע. טסתי במטו...      f
6  אני כבר שנתיים נשוי והשנה אני ואישתי סוף סוף י...      m
7  השנה התחלנו שיפוץ בדירה שלנו בתל אביב. הדירה ה...      f

(753, 2)

In [5]:
df_test.head(3)
df_test.shape

test_example_id                                              story
0                0  כל קיץ אני והמשפחה נוסעים לארצות הברית לוס אנג...
1                1  הגעתי לשירות המדינה אחרי שנתיים כפעיל בתנועת "...
2                2  אחת האהבות הגדולות שלי אלו הכלבים שלי ושל אישת...

(323, 2)

### Your implementation:
Write your code solution in the following code-cells

In [9]:
# changing special chars , notes and numbers

def regOnDate(df):
    processed_df = df.copy() 
    for i in processed_df.index:
        processed_df.iloc[i,0] = re.sub(r'\W',' ',str(processed_df.iloc[i,0])) 
        processed_df.iloc[i,0] = re.sub(r'\d',' ',str(processed_df.iloc[i,0])
        processed_df.iloc[i,0] = re.sub(r'\s\w\s',' ',processed_df.iloc[i,0]) 
    return processed_df['story']

In [10]:
# cleaning the train data 

def my_cleaning():
    df_train['story'] = df_train['story'].apply(lambda x: x.lower())
    df_test['story'] = df_test['story'].apply(lambda x: x.lower())
    df_train['story'] = regOnDate(df_train)
    df_test['story'] = regOnDate(df_test)
    X_train, X_test, y_train, y_test = train_test_split(df_train['story'], df_train['gender'], test_size=0.2, random_state=42)
    return X_train, X_test, y_train, y_test

In [11]:

def f1Avg(y_test, y_pred):
    
    conf_mat = metrics.confusion_matrix(y_test, y_pred) 

    TP = conf_mat[0][0]
    FP = conf_mat[0][1]
    FN = conf_mat[1][0]
    TN = conf_mat[1][1]
    
    precision_f = TN / (TN + FN)  # male precision
    recall_f = TN / (TN + FP)

    precision_m = TP / (TP + FP) # female precision
    recall_m = TP / (TP + FN)


    F1_m = 2 * (precision_f * recall_f) / (precision_f + recall_f)
    F1_f = 2 * (precision_m * recall_m) / (precision_m + recall_m)

    F1_avg = (F1_f + F1_m) / 2
    
    return F1_avg


In [12]:

x_train,x_test,y_train,y_test = my_cleaning()
best_score = -1
good_clf = None


### Finding Best MODEL:

In the following cells , we will try several models , and compare them in order to find the best model

We will use a variety of vectorizers , models - and different values for each model

## KNeighborsClassifier :

In [13]:

max_features = [1500, 1750, 2000]
n_neighbors = [3, 5, 7,9,11]

text_clf_KNN = Pipeline([
    ('vect', CountVectorizer(ngram_range=(1, 6))),
    ('clf', KNeighborsClassifier())
])

param_grid = {
    'vect__max_features': max_features,
    'clf__n_neighbors': n_neighbors
}
grid_search = GridSearchCV(text_clf_KNN, param_grid, cv=10, scoring='f1_macro')

grid_search.fit(x_train, y_train)

best_model = grid_search.best_estimator_
best_model.fit(x_train, y_train)
predicted_test = best_model.predict(x_test)
f1Avg(y_test, predicted_test)
print("Best hyperparameters: %s" % grid_search.best_params_)
print("Cross-validation score: %0.2f" % grid_search.best_score_)


GridSearchCV(cv=10,
             estimator=Pipeline(steps=[('vect',
                                        CountVectorizer(ngram_range=(1, 6))),
                                       ('clf', KNeighborsClassifier())]),
             param_grid={'clf__n_neighbors': [3, 5, 7, 9, 11],
                         'vect__max_features': [1500, 1750, 2000]},
             scoring='f1_macro')

Pipeline(steps=[('vect',
                 CountVectorizer(max_features=2000, ngram_range=(1, 6))),
                ('clf', KNeighborsClassifier(n_neighbors=3))])

0.4957897793263647

Best hyperparameters: {'clf__n_neighbors': 3, 'vect__max_features': 2000}
Cross-validation score: 0.53


##  DecisionTreeClassifier :

In [14]:
max_features = [1500, 1750, 2000]
max_depth = [10, 20, 30, 32]

text_clf_DT = Pipeline([
    ('vect', TfidfVectorizer(min_df=5,ngram_range=(1, 6))),
    ('clf', DecisionTreeClassifier(random_state=42))
])
grid_search_DT = GridSearchCV(text_clf_DT, 
                              {'vect__max_features': max_features, 
                               'clf__max_depth': max_depth}, 
                              cv=10, 
                              scoring='f1_macro')

grid_search_DT.fit(x_train, y_train)
best_text_clf_DT = grid_search_DT.best_estimator_
best_text_clf_DT.fit(x_train, y_train)
predicted_test = best_text_clf_DT.predict(x_test)
f1Avg(y_test, predicted_test)
print("Best hyperparameters: %s" % grid_search_DT.best_params_)
print("Cross-validation score: %0.2f" % grid_search_DT.best_score_)

GridSearchCV(cv=10,
             estimator=Pipeline(steps=[('vect',
                                        TfidfVectorizer(min_df=5,
                                                        ngram_range=(1, 6))),
                                       ('clf',
                                        DecisionTreeClassifier(random_state=42))]),
             param_grid={'clf__max_depth': [10, 20, 30, 32],
                         'vect__max_features': [1500, 1750, 2000]},
             scoring='f1_macro')

Pipeline(steps=[('vect',
                 TfidfVectorizer(max_features=1500, min_df=5,
                                 ngram_range=(1, 6))),
                ('clf', DecisionTreeClassifier(max_depth=10, random_state=42))])

0.6393916540975364

Best hyperparameters: {'clf__max_depth': 10, 'vect__max_features': 1500}
Cross-validation score: 0.60


#### We'll add more parameters and try again :

## Perceptron :

In [15]:
pipe = Pipeline([
    ('vect', CountVectorizer()),
    ('classifier', Perceptron())
])

# Defining the hyperparameters to tune

param_grid = {
    'classifier__penalty': ['l1', 'l2', 'elasticnet'],
    'classifier__alpha': [0.0001, 0.001, 0.01],
    'classifier__max_iter': [1000, 5000, 10000]
}

# Using GridSearchCV to find the best hyperparameters

grid_search = GridSearchCV(pipe, param_grid=param_grid, cv=10, scoring='f1_macro')
grid_search.fit(x_train, y_train)

# Calculating the f1-score on the test-set using the best model

y_pred = grid_search.predict(x_test)
f1 = f1_score(y_test, y_pred, average='macro')
print("F1-score:", f1)

# Calculating the average f1-score using cross-validation

scores = cross_val_score(grid_search, x_train,y_train, cv=10, scoring='f1_macro')
avg_f1 = sum(scores) / len(scores)
print("Average F1-score:", avg_f1)

GridSearchCV(cv=10,
             estimator=Pipeline(steps=[('vect', CountVectorizer()),
                                       ('classifier', Perceptron())]),
             param_grid={'classifier__alpha': [0.0001, 0.001, 0.01],
                         'classifier__max_iter': [1000, 5000, 10000],
                         'classifier__penalty': ['l1', 'l2', 'elasticnet']},
             scoring='f1_macro')

F1-score: 0.6524849163401132
Average F1-score: 0.6113655326264817


## SGDClassifier :

In [16]:
# Defining the hyperparameter grid to be searched over

params = {'vect__min_df': [1, 5],
          'vect__ngram_range': [(1,1), (1,2), (1,3), (1,4), (1,5)],
          'clf__alpha': [0.0001, 0.001, 0.01],
          'clf__penalty': ['l1', 'l2', 'elasticnet'],
          'clf__max_iter': [10, 50, 100],
          'clf__eta0': [0.1, 0.01, 0.001]}

# Defining the pipeline

text_clf = Pipeline([
    ('vect', CountVectorizer()),
    ('clf', SGDClassifier())
])


# Fitting the grid search with cross-validation

grid_search = GridSearchCV(text_clf, params, cv=10, scoring='f1_macro')
grid_search.fit(x_train, y_train)

# Getting the best estimator and printing it's hyperparameters and cross-validation score:

best_clf = grid_search.best_estimator_
f1Avg(y_test, y_pred)
print("Best hyperparameters: %s" % grid_search.best_params_)
print("Cross-validation score: %0.2f" % grid_search.best_score_)

GridSearchCV(cv=10,
             estimator=Pipeline(steps=[('vect', CountVectorizer()),
                                       ('clf', SGDClassifier())]),
             param_grid={'clf__alpha': [0.0001, 0.001, 0.01],
                         'clf__eta0': [0.1, 0.01, 0.001],
                         'clf__max_iter': [10, 50, 100],
                         'clf__penalty': ['l1', 'l2', 'elasticnet'],
                         'vect__min_df': [1, 5],
                         'vect__ngram_range': [(1, 1), (1, 2), (1, 3), (1, 4),
                                               (1, 5)]},
             scoring='f1_macro')

0.6524849163401132

Best hyperparameters: {'clf__alpha': 0.0001, 'clf__eta0': 0.01, 'clf__max_iter': 100, 'clf__penalty': 'l1', 'vect__min_df': 5, 'vect__ngram_range': (1, 4)}
Cross-validation score: 0.68


#### While the f1 score and Cross-Val score got better , we can still improve them :

## LinearSVC :

In [17]:
# Defining the pipeline

pipeline = Pipeline([
    ('tfidf', TfidfVectorizer()),
    ('clf', LinearSVC())
])

# Defining the hyperparameters to be tuned:

parameters = {
    'tfidf__ngram_range': [(1, 1), (1, 2)],
    'tfidf__use_idf': (True, False),
    'clf__C': [0.1, 1, 10]
}

# Using GridSearchCV to find the best hyperparameters:

grid_search = GridSearchCV(pipeline, parameters, cv=10, n_jobs=-1)
grid_search.fit(x_train, y_train)


best_text_clf_DT = grid_search.best_estimator_
best_text_clf_DT.fit(x_train, y_train)
predicted_test = grid_search.predict(x_test)
f1Avg(y_test, predicted_test)
print("Best hyperparameters: %s" % grid_search.best_params_)
print("Cross-validation score: %0.2f" % grid_search.best_score_)

GridSearchCV(cv=10,
             estimator=Pipeline(steps=[('tfidf', TfidfVectorizer()),
                                       ('clf', LinearSVC())]),
             n_jobs=-1,
             param_grid={'clf__C': [0.1, 1, 10],
                         'tfidf__ngram_range': [(1, 1), (1, 2)],
                         'tfidf__use_idf': (True, False)})

Pipeline(steps=[('tfidf', TfidfVectorizer()), ('clf', LinearSVC(C=10))])

0.6346266166040885

Best hyperparameters: {'clf__C': 10, 'tfidf__ngram_range': (1, 1), 'tfidf__use_idf': True}
Cross-validation score: 0.80


### F1 score is still low at 0.63 
### but , Cross-Val score is now at 0.80

##### In the following cell , we will figure out , which parameteres lead to the best mean scores:

## MultinomialNB :

In [18]:
max_features = [1600, 1700, 1800]
alpha = [0.1, 1, 10]

best_score = 0
best_params = {}

for mf in max_features:
    for a in alpha:
        text_clf_NB = Pipeline([
            ('vect', CountVectorizer(max_features=mf, ngram_range=(1, 6))),
            ('clf', MultinomialNB(alpha=a))
        ])

        scores = cross_val_score(text_clf_NB, x_train, y_train, cv=10, scoring='f1_macro')

        score = np.mean(scores)

        if score > best_score:
            best_score = score
            best_params = {'max_features': mf, 'alpha': a}

text_clf_NB = Pipeline([
    ('vect', CountVectorizer(max_features=best_params['max_features'], ngram_range=(1, 6))),
    ('clf', MultinomialNB(alpha=best_params['alpha']))
])

#fitting the model and predicting results:

text_clf_NB.fit(x_train, y_train)
predicted_test = text_clf_NB.predict(x_test)
f1Avg(y_test, predicted_test)

print("Best hyperparameters: %s" % best_params)
print("Cross-validation score: %0.2f" % best_score)

Pipeline(steps=[('vect',
                 CountVectorizer(max_features=1700, ngram_range=(1, 6))),
                ('clf', MultinomialNB(alpha=1))])

0.7026992696093999

Best hyperparameters: {'max_features': 1700, 'alpha': 1}
Cross-validation score: 0.67


##### Now the f1 score is much better , but we need to improve the Cross-Val score:

## MLPClassifier :

In [20]:
# Defining the pipeline

pipeline = Pipeline([
    ('tfidf', TfidfVectorizer()),
    ('clf', MLPClassifier())
])

# Defining the hyperparameters for a final improved model:

parameters = {
    'tfidf__ngram_range': [(1, 1), (1, 2)],
    'tfidf__use_idf': (True, False),
    'clf__hidden_layer_sizes': [(50,), (100,), (50, 50), (100, 100)],
    'clf__alpha': [0.0001, 0.001, 0.01]
}

# Using GridSearchCV to find the best hyperparameters:

grid_search = GridSearchCV(pipeline, parameters, cv=10, n_jobs=-1)
grid_search.fit(x_train, y_train)

# fitting the best estimators , parameters and predicting using GSCV

best_text_clf_DT = grid_search.best_estimator_

best_text_clf_DT.fit(x_train, y_train)

predicted_test = grid_search.predict(x_test)

f1Avg(y_test, predicted_test)

print("Best hyperparameters: %s" % grid_search.best_params_)
print("Cross-validation score: %0.2f" % grid_search.best_score_)

GridSearchCV(cv=10,
             estimator=Pipeline(steps=[('tfidf', TfidfVectorizer()),
                                       ('clf', MLPClassifier())]),
             n_jobs=-1,
             param_grid={'clf__alpha': [0.0001, 0.001, 0.01],
                         'clf__hidden_layer_sizes': [(50,), (100,), (50, 50),
                                                     (100, 100)],
                         'tfidf__ngram_range': [(1, 1), (1, 2)],
                         'tfidf__use_idf': (True, False)})

Pipeline(steps=[('tfidf', TfidfVectorizer(use_idf=False)),
                ('clf',
                 MLPClassifier(alpha=0.01, hidden_layer_sizes=(100, 100)))])

0.7361559139784947

Best hyperparameters: {'clf__alpha': 0.01, 'clf__hidden_layer_sizes': (100, 100), 'tfidf__ngram_range': (1, 1), 'tfidf__use_idf': False}
Cross-validation score: 0.80


### Final Result :
#### Best F1 score : 0.736
#### Best Cross-Val score : 0.80
#### Best model : MLPClassifier

##### Now , we will predict on the test set :

In [258]:
X_test = df_test['story']
predDfTest = best_text_clf_DT.predict(X_test) # predicting if the author is male or female
test = df_test.copy()

for i in range(0,len(test['test_example_id'])):
    test['test_example_id'][i] = predDfTest[i] #inserting the predictions into the data itself

###### In the tabs bellow , you can see the dataset's 5 first and last rows - of the original 
###### and the predicted data: 

In [259]:
df_predicted = pd.DataFrame(columns = ['test_example_id','predicted_category'])

df_predicted['test_example_id'] = df_test['test_example_id']
df_predicted['predicted_category'] = test['test_example_id']
test
df_predicted

test_example_id                                              story
0                 m  כל קיץ אני והמשפחה נוסעים לארצות הברית לוס אנג...
1                 m  הגעתי לשירות המדינה אחרי שנתיים כפעיל בתנועת "...
2                 m  אחת האהבות הגדולות שלי אלו הכלבים שלי ושל אישת...
3                 m  רגע הגיוס לצבא היה הרגע הכי משמעותי עבורי, אני...
4                 f  אני הגעתי לברזיל ישר מקולומביה וגם אני עשיתי ע...
..              ...                                                ...
318               m  בשנה האחרונה הרגשתי די תקוע בעבודה, השגרה הפכה...
319               m  אני ואילן חברים טובים מזה 20 שנה תמיד חלמנו לפ...
320               m  מידי יום שישי אני נוהג לנסוע בתחבורה ציבורית ס...
321               m  לפני מספר חודשים, בשיא התחלואה של הגל השני, עמ...
322               m  היום בו דיווחתי על גניבה של האוטו שלי. בוקר אח...

[323 rows x 2 columns]

test_example_id predicted_category
0                  0                  m
1                  1                  m
2                  2                  m
3                  3                  m
4                  4                  f
..               ...                ...
318              318                  m
319              319                  m
320              320                  m
321              321                  m
322              322                  m

[323 rows x 2 columns]

### Save output to csv (optional)
After you're done save your output to the 'classification_results.csv' csv file.<br/>
We assume that the dataframe with your results contain the following columns:
* column 1 (left column): 'test_example_id'  - the same id associated to each of the test stories to be predicted.
* column 2 (right column): 'predicted_category' - the predicted gender value for each of the associated story. 

Assuming your predicted values are in the `df_predicted` dataframe, you should save you're results as following:

In [ ]:
df_predicted.to_csv('classification_results.csv',index=False)